geode-core/src/test/java/org/apache/geode/internal/cache/PartitionedRegionTest.java

```diff
@@ -573,6 +578,50 @@ public void transactionThrowsTransactionDataRebalancedExceptionIfIsAForceReattem
         .hasMessage(PartitionedRegion.DATA_MOVED_BY_REBALANCE).hasCause(exception);
   }
 
+  @Test
+  public void failuresSavedIfFetchKeysThrows() throws Exception {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    VersionedObjectList values = mock(VersionedObjectList.class);
+    ServerConnection serverConnection = mock(ServerConnection.class);
+    Set<Integer> failures = mock(Set.class);
+    InternalDistributedMember member = mock(InternalDistributedMember.class);
+    Set<Integer> buckets = new HashSet<>();
+    buckets.add(1);
+    doThrow(new ForceReattemptException("")).when(spyPartitionedRegion).getFetchKeysResponse(member,
+        1);
+
+    spyPartitionedRegion.fetchKeysAndValues(values, serverConnection, failures, member, null,
+        buckets);
+
+    verify(failures).add(1);
+    verify(spyPartitionedRegion, never()).getValuesForKeys(values, serverConnection, null);
+  }
+
+  @Test
+  public void fetchKeysAndValuesInvokesGetValuesForKeys() throws Exception {
+    PartitionedRegion spyPartitionedRegion = spy(partitionedRegion);
+
+    VersionedObjectList values = mock(VersionedObjectList.class);
+    ServerConnection serverConnection = mock(ServerConnection.class);
+    Set<Integer> failures = mock(Set.class);
```

Comment:
Here you should use a real Set instead of a mock.

Reply:
Is that a particular reason set should not be mocked? It is not being iterated through, just used as a placeholder. I would like to know the reason for the requested change so that I may gain some new knowledge.

Reply:
It's part of the "don't mock what you don't own" principle in TDD, explained [here](https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own) and [here](https://blog.codecentric.de/en/2018/03/mock-what-when-how/). Also, as discussed [here](https://github.com/mockito/mockito/wiki/How-to-write-good-tests), if you can avoid mocking a class, as in this case, then that's probably a good thing.

Reply:
Thanks for the explanation. Change has been made.